File: notegoat.pl
use strict;
use warnings;
use File::Copy qw(move);

# The name of the file where the notes will be stored in SQL format
my $SQL_FILENAME = 'notes.sql';

sub load_existing_notes {
    # Load existing notes from the SQL file into a hash
    my %notes;
    if (open my $file, '<', $SQL_FILENAME) {
        while (my $line = <$file>) {
            # Use regex to find note names in the INSERT INTO statements
            if ($line =~ /INSERT INTO notes \(name, address, phone, email, note\) VALUES \('([^']+)'/) {
                $notes{$1} = 1;  # Store the note name in the hash
            }
        }
        close $file;
    }
    return %notes;  # Return the hash of note names
}

sub write_sql_header {
    # Write the SQL table header to the file if it does not already exist
    if (!-e $SQL_FILENAME) {
        open my $file, '>', $SQL_FILENAME or die "Could not open file '$SQL_FILENAME': $!";
        print $file "CREATE TABLE IF NOT EXISTS notes (\n";
        print $file "    id INTEGER PRIMARY KEY AUTOINCREMENT,\n";
        print $file "    name TEXT UNIQUE,\n";
        print $file "    address TEXT,\n";
        print $file "    phone TEXT,\n";
        print $file "    email TEXT,\n";
        print $file "    note TEXT\n";
        print $file ");\n\n";
        close $file;
    }
}

sub add_note {
    # Add a new note to the SQL file
    my %existing_notes = load_existing_notes();

    print "Name: ";
    chomp(my $name = <STDIN>);
    if (exists $existing_notes{$name}) {
        print "A note with the name '$name' already exists.\n";
        return;
    }

    print "Address: ";
    chomp(my $address = <STDIN>);

    my $phone;
    while (1) {
        print "Phone number: ";
        chomp($phone = <STDIN>);
        if ($phone =~ /^[0-9\s\-]+$/) {
            last;
        } else {
            print "Invalid phone number. Please enter a valid phone number (digits, spaces, and hyphens allowed).\n";
        }
    }

    my $email;
    while (1) {
        print "Email address: ";
        chomp($email = <STDIN>);
        if ($email =~ /@/) {
            last;
        } else {
            print "Missing @. Please enter a valid email address.\n";
        }
    }

    print "Note: ";
    chomp(my $note = <STDIN>);

    # Append the new note to the SQL file
    open my $file, '>>', $SQL_FILENAME or die "Could not open file '$SQL_FILENAME': $!";
    print $file "INSERT INTO notes (name, address, phone, email, note) VALUES ('$name', '$address', '$phone', '$email', '$note');\n";
    close $file;

    print "Note added successfully.\n";
}

sub delete_note {
    # Delete a note from the SQL file based on the note name provided by the user
    print "Enter the name of the note to delete: ";
    chomp(my $search_name = <STDIN>);

    my $tempfile = 'temp.sql';
    my $deleted = 0;

    open my $infile, '<', $SQL_FILENAME or die "Could not open file '$SQL_FILENAME': $!";
    open my $outfile, '>', $tempfile or die "Could not open file '$tempfile': $!";
    
    while (my $line = <$infile>) {
        if ($line =~ /INSERT INTO notes \(name, address, phone, email, note\) VALUES \('$search_name',/) {
            $deleted = 1;  # Mark the note as deleted
            next;
        }
        print $outfile $line;  # Copy all other lines to the temp file
    }

    close $infile;
    close $outfile;

    # Replace the original file with the updated file
    move($tempfile, $SQL_FILENAME) or die "Could not move file '$tempfile' to '$SQL_FILENAME': $!";

    if ($deleted) {
        print "Note deleted successfully.\n";
    } else {
        print "No note with the name '$search_name' found.\n";
    }
}

sub search_note {
    # Search for a note in the SQL file by the note name provided by the user
    print "Enter the name of the note to search: ";
    chomp(my $search_name = <STDIN>);

    my $found = 0;

    open my $file, '<', $SQL_FILENAME or die "Could not open file '$SQL_FILENAME': $!";
    while (my $line = <$file>) {
        if ($line =~ /INSERT INTO notes \(name, address, phone, email, note\) VALUES \('$search_name', '([^']*)', '([^']*)', '([^']*)', '([^']*)'\);/) {
            my ($address, $phone, $email, $note) = ($1, $2, $3, $4);
            print "Name: $search_name\n";
            print "Address: $address\n";
            print "Phone: $phone\n";
            print "Email: $email\n";
            print "Note: $note\n";
            print "---------------\n";
            $found = 1;
            last;
        }
    }
    close $file;

    if (!$found) {
        print "No note with the name '$search_name' found.\n";
    }
}

sub edit_note {
    # Edit an existing note by first deleting the old note and then adding a new note with the same name
    print "Enter the name of the note to edit: ";
    chomp(my $search_name = <STDIN>);

    my %existing_notes = load_existing_notes();
    if (!exists $existing_notes{$search_name}) {
        print "No note with the name '$search_name' found.\n";
        return;
    }

    delete_note_by_name($search_name);  # Delete the old note
    add_note();  # Add a new note with the same name
    print "Note edited successfully.\n";
}

sub delete_note_by_name {
    # Helper function to delete a note by name without user interaction
    my ($search_name) = @_;

    my $tempfile = 'temp.sql';

    open my $infile, '<', $SQL_FILENAME or die "Could not open file '$SQL_FILENAME': $!";
    open my $outfile, '>', $tempfile or die "Could not open file '$tempfile': $!";
    
    while (my $line = <$infile>) {
        if ($line =~ /INSERT INTO notes \(name, address, phone, email, note\) VALUES \('$search_name',/) {
            next;  # Skip the line that matches the note to delete
        }
        print $outfile $line;  # Copy all other lines to the temp file
    }

    close $infile;
    close $outfile;

    # Replace the original file with the updated file
    move($tempfile, $SQL_FILENAME) or die "Could not move file '$tempfile' to '$SQL_FILENAME': $!";
}

sub main {
    # Main function to run the note-taking application
    write_sql_header();  # Ensure the SQL file has the table schema
    while (1) {
        print "1. Add Note\n";
        print "2. Delete Note\n";
        print "3. Search Note\n";
        print "4. Edit Note\n";
        print "5. Exit\n";
        print "Enter your choice: ";
        chomp(my $choice = <STDIN>);

        if ($choice == 1) {
            add_note();
        } elsif ($choice == 2) {
            delete_note();
        } elsif ($choice == 3) {
            search_note();
        } elsif ($choice == 4) {
            edit_note();
        } elsif ($choice == 5) {
            last;
        } else {
            print "Invalid choice\n";
        }
    }
}

main();  # Run the main function to start the application

File: notegoat.py
import re
import os

# The name of the file where the notes will be stored in SQL format
SQL_FILENAME = 'notes.sql'

def load_existing_notes():
    """
    Loads the existing notes from the SQL file.
    Returns a set of note names that are already in the file.
    """
    notes = set()
    try:
        with open(SQL_FILENAME, 'r') as file:
            for line in file:
                # Use regex to find note names in the INSERT INTO statements
                match = re.match(r"INSERT INTO notes \(name, address, phone, email, note\) VALUES \('([^']+)'", line)
                if match:
                    notes.add(match.group(1))  # Add the note name to the set
    except FileNotFoundError:
        pass  # If the file doesn't exist, return an empty set
    return notes

def write_sql_header():
    """
    Writes the SQL table header to the file if it does not already exist.
    This creates the table schema for storing notes.
    """
    if not os.path.exists(SQL_FILENAME):
        with open(SQL_FILENAME, 'w') as file:
            file.write("CREATE TABLE IF NOT EXISTS notes (\n")
            file.write("    id INTEGER PRIMARY KEY AUTOINCREMENT,\n")
            file.write("    name TEXT UNIQUE,\n")
            file.write("    address TEXT,\n")
            file.write("    phone TEXT,\n")
            file.write("    email TEXT,\n")
            file.write("    note TEXT\n")
            file.write(");\n\n")

def add_note():
    """
    Adds a new note to the SQL file.
    Ensures the note name is unique and collects other note details from the user.
    """
    existing_notes = load_existing_notes()

    name = input("Name: ").strip()
    if name in existing_notes:
        print(f"A note with the name '{name}' already exists.")
        return

    address = input("Address: ").strip()

    # Regex pattern to validate phone number
    phone_pattern = re.compile(r'^[0-9\s\-]+$')
    while True:
        phone = input("Phone number: ").strip()
        if phone_pattern.match(phone):
            break
        else:
            print("Invalid phone number. Please enter a valid phone number (digits, spaces, and hyphens allowed).")

    while True:
        email = input("Email address: ").strip()
        if '@' in email:
            break
        else:
            print("Missing @. Please enter a valid email address.")

    note = input("Note: ").strip()

    # Write the new note to the SQL file
    with open(SQL_FILENAME, 'a') as file:
        sql_command = f"INSERT INTO notes (name, address, phone, email, note) VALUES ('{name}', '{address}', '{phone}', '{email}', '{note}');\n"
        file.write(sql_command)
    print("Note added successfully.")

def delete_note():
    """
    Deletes a note from the SQL file based on the note name provided by the user.
    """
    search_name = input("Enter the name of the note to delete: ").strip()

    tempfile = 'temp.sql'
    deleted = False
    with open(SQL_FILENAME, 'r') as infile, open(tempfile, 'w') as outfile:
        for line in infile:
            if f"INSERT INTO notes (name, address, phone, email, note) VALUES ('{search_name}'," in line:
                deleted = True  # Mark as deleted
            else:
                outfile.write(line)  # Copy all other lines to the temp file

    os.replace(tempfile, SQL_FILENAME)  # Replace the original file with the updated file

    if deleted:
        print("Note deleted successfully.")
    else:
        print(f"No note with the name '{search_name}' found.")

def search_note():
    """
    Searches for a note in the SQL file by the note name provided by the user.
    Displays the note details if found.
    """
    search_name = input("Enter the name of the note to search: ").strip()

    found = False
    with open(SQL_FILENAME, 'r') as file:
        for line in file:
            if f"INSERT INTO notes (name, address, phone, email, note) VALUES ('{search_name}'," in line:
                found = True
                # Extract the values from the SQL command
                match = re.search(r"INSERT INTO notes \(name, address, phone, email, note\) VALUES \('([^']*)', '([^']*)', '([^']*)', '([^']*)', '([^']*)'\);", line)
                if match:
                    name, address, phone, email, note = match.groups()
                    print(f"Name: {name}")
                    print(f"Address: {address}")
                    print(f"Phone: {phone}")
                    print(f"Email: {email}")
                    print(f"Note: {note}")
                    print(f"---------------")
                break

    if not found:
        print(f"No note with the name '{search_name}' found.")

def edit_note():
    """
    Edits an existing note by first deleting the old note and then adding a new note with the same name.
    """
    search_name = input("Enter the name of the note to edit: ").strip()

    existing_notes = load_existing_notes()
    if search_name not in existing_notes:
        print(f"No note with the name '{search_name}' found.")
        return

    delete_note_by_name(search_name)  # Delete the old note
    add_note()  # Add a new note with the same name
    print("Note edited successfully.")

def delete_note_by_name(search_name):
    """
    Helper function to delete a note by name without user interaction.
    Used internally by edit_note.
    """
    tempfile = 'temp.sql'
    with open(SQL_FILENAME, 'r') as infile, open(tempfile, 'w') as outfile:
        for line in infile:
            if f"INSERT INTO notes (name, address, phone, email, note) VALUES ('{search_name}'," in line:
                continue  # Skip the line that matches the note to delete
            outfile.write(line)  # Copy all other lines to the temp file

    os.replace(tempfile, SQL_FILENAME)  # Replace the original file with the updated file

def main():
    """
    Main function to run the note-taking application.
    Displays a menu to the user for different actions: add, delete, search, edit, and exit.
    """
    write_sql_header()  # Ensure the SQL file has the table schema
    while True:
        print("1. Add Note")
        print("2. Delete Note")
        print("3. Search Note")
        print("4. Edit Note")
        print("5. Exit")
        choice = input("Enter your choice: ").strip()

        if choice == '1':
            add_note()
        elif choice == '2':
            delete_note()
        elif choice == '3':
            search_note()
        elif choice == '4':
            edit_note()
        elif choice == '5':
            break
        else:
            print("Invalid choice")

if __name__ == '__main__':
    main()
